feat: basic Sentry error monitoring

bajtos opened the pull request:
See #78

- feat: setup Sentry error monitoring
- feat: add release to Sentry config

Out of scope:
- Report releases and source maps to Sentry
- Report errors from the backend

juliangruber commented on main/index.js:
```diff
@@ -21,6 +22,13 @@ const { ipcMain } = require('electron/main')
 const inTest = (process.env.NODE_ENV === 'test')
 const isDev = !app.isPackaged && !inTest
 
+// Expose additional metadata for Electron preload script
```
Is overwriting `process.env` the best way to pass data to the client? 🤔

juliangruber replied:
Tbh I haven't thought about how to do this in a setup where the renderer doesn't have access to `require('electron')`

bajtos replied:
The preload script is intentionally sandboxed for security reasons. It has access to a very limited set of APIs, see https://www.electronjs.org/docs/latest/tutorial/tutorial-preload

I am open to use a different mechanism for passing data from `main` to `preload`, what do you suggest?

juliangruber replied:
I don't have anything on the top of my mind, I was more curious if there were well known alternatives. Thanks for the context!

juliangruber commented on main/index.js:
```diff
@@ -21,6 +22,13 @@ const { ipcMain } = require('electron/main')
+process.env.IS_PACKAGED = app.isPackaged ? '1' : undefined
+process.env.STATION_VERSION = app.isPackaged
+  ? STATION_VERSION
+  // TODO: can we use Git commit SHA instead?
```
We probably could, but is that a meaningful change for now?

bajtos replied:
It depends on what we want to see in the Sentry dashboard when looking at errors. At the moment, when running the Station via `npm start`, errors are labelled with release `dev-on-0.5.0`. If we think that's good enough, then I can remove this TODO comment.

I consider this as a longer-term idea, not something to work on right now. I think we would need to obtain the Git commit SHA in a sync way (during the first tick of the event loop), which is tricky.

juliangruber replied:
Since we can annotate dev environments in Sentry I don't think we have to remove Sentry from dev - it won't pollute our production date - but at the same time removing it might be the simpler path here?

bajtos replied:
Great discussion!

I see three different environments to cater for:
- Production - packaged builds produced from git tags (releases).
- Testing - automated CI tests run via GitHub Actions. I think we run the app unpackaged.
- Local development

I agree we don't need Sentry for the dev env because we can inspect errors in DevTools.

I think it's potentially useful to collect error reports from our CI runs. Ideally, these errors should be reported by our test suite. Considering the effort required to do that, I feel it's much easier to use Sentry.

juliangruber replied:
Uh that's a great idea! CI can be such a PITA to debug

bajtos replied:
The new version I pushed today, we are reporting all errors to Sentry. Errors from local dev builds can be distinguished via version number ending with `.1-dev`.

Do you agree this is good enough for now, or should I look into ways how to disable error reporting from dev builds?

juliangruber commented on renderer/src/main.tsx:
```diff
@@ -1,9 +1,24 @@
+  // Set tracesSampleRate to 1.0 to capture 100%
+  // of transactions for performance monitoring.
+  // We recommend adjusting this value in production
+  tracesSampleRate: 1.0
```
submitting traces is expensive (and potentially intrusive), because a lot of data will be captured and submitted. Shall we reduce this right away, to say 10%?

bajtos replied:
Good point. Do we actually need tracing at all? Maybe we can start with collecting errors only and wait with adding tracing until we need it?

juliangruber replied:
I'm concerned that we'll never start using traces if we don't at least have _some_ in our dashboard 🤔 But I don't feel strongly about this either, we could just remove them too

bajtos replied:
Makes sense. I'll set the rate to `0.1` then. We can easily tweak this later, based on what we learn from real-world usage.

bajtos commented:
I reworked the Sentry configuration as follows:

- Changed `tracesSampleRate` to `0.1`

- Removed `environment` config. I'd like to configure this in such way that only release build produced for tags are reporting with `production` environment and everything else reports with `development`. It's not straightforward, let's leave it out of the scope of this PR.

- Reworked the version reported to Sentry as follows:
   - Release builds report `x.y.z`, e.g. `0.5.0`
   - Non-release builds from GitHub Actions report `x.y.z.{build-number}`, e.g. `0.5.0.444`
   - Local/dev builds report `x.y.z.1-dev`, e.g. `0.5.0.1-dev`

@juliangruber please review again

juliangruber reviewed (approved):
This is great! Thank you for putting the extra thought into this!
